lib/puppet/provider/keystore_certificate/openssl.rb

```diff
@@ -0,0 +1,95 @@
+Puppet::Type.type(:keystore_certificate).provide(:openssl) do
+  commands :openssl => 'openssl'
+  commands :keytool => 'keytool'
+
+  def create
+    add_certificate
+  end
+
+  def destroy
+    delete_certificate
+  end
+
+  def exists?
+    keystore_content
+  end
+
+  def certificate
+    keystore_fingerprint
+  end
+
+  def certificate=(value)
+    delete_certificate unless keystore_content.nil?
+    add_certificate
+  end
+
+  def fingerprint(file)
+    return unless File.exist?(file)
+
+    openssl('x509', '-sha256', '-noout', '-fingerprint', '-in', file).strip.split('=')[1]
+  end
+
+  private
+
+  def add_certificate
+    Tempfile.open('temp_keystore') do |temp_store|
+      openssl(
+        'pkcs12',
+        '-export',
+        '-in', resource[:certificate],
+        '-inkey', resource[:private_key],
+        '-out', temp_store,
+        '-name', resource[:alias],
+        '-CAfile', resource[:ca],
+        '-password', "file:#{resource[:password_file]}"
+      )
+
+      keytool(
+        '-importkeystore',
+        '-noprompt',
+        '-srckeystore', temp_store,
+        '-srcstorepass:file', resource[:password_file],
+        '-destkeystore', resource[:keystore],
+        '-deststorepass:file', resource[:password_file],
+        '-srcalias', resource[:alias],
+        '-destalias', resource[:alias]
+      )
+    end
+  rescue Puppet::ExecutionFailure => e
+    Puppet.debug("Failed to add certificate to keystore: #{e}")
+    nil
+  end
```

Comment:
```suggestion
  def add_certificate
    Tempfile.open('temp_keystore') do |temp_store|
      openssl(
        'pkcs12',
        '-export',
        '-in', resource[:certificate],
        '-inkey', resource[:private_key],
        '-out', temp_store,
        '-name', resource[:alias],
        '-CAfile', resource[:ca],
        '-password', "file:#{resource[:password_file]}"
      )
      rescue Puppet::ExecutionFailure => e
        Puppet.err("Failed to export certificate to PKCS12: #{e}")
        return
      end

      keytool(
        '-importkeystore',
        '-noprompt',
        '-srckeystore', temp_store,
        '-srcstorepass:file', resource[:password_file],
        '-destkeystore', resource[:keystore],
        '-deststorepass:file', resource[:password_file],
        '-srcalias', resource[:alias],
        '-destalias', resource[:alias]
      )
      rescue Puppet::ExecutionFailure => e
        Puppet.err("Failed to add certificate to keystore: #{e}")
        return
      end
    end
  end
```

This changes:

1. `def add_certificate` was missing `end`

2. Log issues here at error level

3. Move `rescue` blocks into method, and `return`

(I haven't yet tested this)

Reply:
I don't this is quite right. I see now the intended structure was

```
def add_certificate
  Tempfile.open('temp_keystore') do |temp_store|
    openssl(...)
    keytool(...)
  rescue
    ...
  end
end
```

Reply:
I can make the change to error level. The two operations are "atomic" in their execution given the Temp file that is being created. The code gets kinda of ugly attempting to rescue multiple places and ensure the Temp file is cleaned up in each and every case. The failure being caught for either invocation would trigger the rescue and printing useful information to the user. That is why I chose not to split it up.